tempesta_fw/avx_routines.h

```diff
@@ -0,0 +1,152 @@
+#ifndef AVX_ROUTINES_H
+#define AVX_ROUTINES_H
+
+#pragma GCC push_options
+#pragma GCC target("sse","sse2","sse3","ssse3","sse4.1","avx","avx2")
+#define _MM_MALLOC_H_INCLUDED
+#include <x86intrin.h>
+#undef  _MM_MALLOC_H_INCLUDED
+#pragma GCC pop_options
+
+extern const unsigned char __c_avx_data[];
+extern const unsigned char __c_avx_header_names[];
+
+#define __header_charset _mm256_loadu2_m128i((const __m128i*)__c_avx_data, (const __m128i*)__c_avx_data)
+#define __constants _mm256_loadu2_m128i((const __m128i*)__c_avx_data+16, (const __m128i*)__c_avx_data+16)
+#define _mm256_loadu(p) _mm256_lddqu_si256((const __m256*)p)
+#define _mm256_load(p)  _mm256_load_si256((const __m256*)p)
+
+#define __DEFINE_AVX_VARIABLES \
+    __m256 HEADER_CHARSET, __c_avx_constants1, __c_avx_constants2; \
+    unsigned char __avx_realign[64] __attribute__((aligned(64))); \
+    do {\
+        __m256 tmp = _mm256_load(__c_avx_data);\
+        __m256 tmp2 = _mm256_load(__c_avx_data+32);\
+        HEADER_CHARSET = _mm256_permute2f128_si256(tmp, tmp, 0x00);\
+        __m256 tmp3 = _mm256_permute4x64_epi64(tmp, 0xEE);\
+        tmp2 = _mm256_unpacklo_epi32(tmp3, tmp3);\
+        __c_avx_constants1 = tmp3;\
+        __c_avx_constants2 = tmp2;\
+    }while(0);
```

Comment:
As I see it, `HEADER_CHARSET, __c_avx_constants1, __c_avx_constants2` can be precalculated.